src/envoy/mixer/mixer_control.cc

```diff
@@ -70,10 +72,13 @@ HttpMixerControl::HttpMixerControl(const HttpMixerConfig& mixer_config,
   ::istio::mixer_control::http::Controller::Options options(
       mixer_config.http_config, mixer_config.legacy_quotas);
 
-  CreateEnvironment(cm, dispatcher, random, &options.env);
+  check_cluster_ = mixer_config.check_cluster;
+  report_cluster_ = mixer_config.report_cluster;
 
-  controller_ = ::istio::mixer_control::http::Controller::Create(options);
+  CreateEnvironment(cm, dispatcher, random, check_cluster_, report_cluster_,
```

Comment:
I think you will not need check_cluster_ and report_cluster_ in MixerControl object if you don't pass reference to CreateEnvironment().    Just pass std::string and make a copy

Reply:
passing const reference is fine as long as the lamda in GetFunc captures cluster_name as copy, so just remove check_cluster_ and report_cluster_ is good.

Reply:
Oh I needed to persist these variables in mixer control object because there is a GetFunc call in http_filter.cc that does `mixer_control_->...` ..
Ideally, i would have liked to stuff them into the options object, but that happens to be defined in mixerclient (spreading to too many repos).

Does that your point @qiwzhang ?

Reply:
I see. That one is per-route Check transport function.  My suggestion is:

in both HTTPMixerControl and TcpMixerControl store a const reference to MixerConfig

const MixerConfig& config_;

Expose it as

const MxerConfig& config() const { return config_; }

use the check_cluster_ as

 config()->check_cluster()

It is better than store anothery copy in MixerControl

Reply:
Well, since the PR is merged.  I will clean it up later.

Reply:
Here is my PR https://github.com/istio/proxy/pull/1001
to clean this up